src/freak.jl

```diff
@@ -0,0 +1,105 @@
+type FREAK{S, T, O, N} <: DescriptorParams
+    pattern_scale::Float64
+    pattern_table::Array{S, N}
+    smoothing_table::Array{T, N}
+    orientation_weights::Array{O}
+end
+
+typealias OrientationPair Vector{Int}
+typealias SamplePair Vector{Float64}
+
+function FREAK(; pattern_scale::Float64 = 22.0)
+    pattern_table, smoothing_table = _freak_tables(pattern_scale)
+    orientation_weights = OrientationPair[]
+    for o in freak_orientation_sampling_pattern
+        offset_1 = pattern_table[1][o[1]]
+        offset_2 = pattern_table[1][o[2]]
+        dy, dx = offset_1 - offset_2
+        norm = (dx ^ 2 + dy ^ 2)
+        push!(orientation_weights, OrientationPair([round(Int, dy * 4096 / norm), round(Int, dx * 4096 / norm)]))
+    end
+    FREAK(pattern_scale, pattern_table, smoothing_table, orientation_weights)
+end
+
+function _freak_mean_intensity{T<:Gray}(int_img::AbstractArray{T, 2}, keypoint::Keypoint, offset::SamplePair, sigma::Float64)
+    y = k[1] + offset[1]
+    x = k[2] + offset[2]
+    if sigma < 0.5
+        tl = bilinear_interpolation(int_img, y - sigma, x - sigma)
+        tr = bilinear_interpolation(int_img, y - sigma, x + sigma)
+        bl = bilinear_interpolation(int_img, y + sigma, x)
+        br = bilinear_interpolation(int_img, y + sigma, x + sigma)
+        return (br + tl - tr - bl) / (4 * (sigma ^ 2))
+    end
+    xs = round(Int, x - sigma)
+    ys = round(Int, y - sigma)
+    xst = round(Int, x + sigma)
+    yst = round(Int, y + sigma)
+    intensity = boxdiff(int_img, ys:yst, xs:xst)
+    intensity / ((xst - xs) * (yst - ys))
+end
+
+function _freak_orientation{T<:Gray}(int_img::AbstractArray{T, 2}, keypoint::Keypoint, pattern::Array{SamplePair}, 
+                                        orientation_weights::Array{OrientationPair}, sigmas::Array{Float64})
+    direction_sum_y = 0.0
+    direction_sum_x = 0.0
+    for (i, o) in enumerate(freak_orientation_sampling_pattern)
+        offset_1 = pattern[o[1]]
+        offset_2 = pattern[o[2]]
+        intensity_diff = _freak_mean_intensity(int_img, k, offset_1, sigmas[o[1]]) - _freak_mean_intensity(int_img, k, offset_2, sigmas[o[2]])
+        direction_sum_y += orientation_weights[i] * intensity_diff / 4096
+        direction_sum_x += orientation_weights[i] * intensity_diff / 4096
+    end
+    angle = atan2(direction_sum_y, direction_sum_x)
+    scaled_angle = round(Int, (angle + pi) * freak_orientation_steps / (2 * pi))
+    scaled_angle
+end
+
+function _freak_tables(pattern_scale::Float64)
+    pattern_table = Array{SamplePair}[]
```

Comment:
In case you're confused, again remember that function signatures are totally different from creation in terms of whether it matters. (Functions get specialized on the types that actually get passed in, so are automatically using leaf types if that's what you pass in. Type declarations and created containers do exactly what you tell them.)